File: 2015/04/solution.py
import hashlib
from itertools import count

secret_key = input()
p1 = next(
    i
    for i in count(1)
    if hashlib.md5(f"{secret_key}{i}".encode()).hexdigest()[:5] == "00000"
)
p2 = next(
    i
    for i in count(p1)
    if hashlib.md5(f"{secret_key}{i}".encode()).hexdigest()[:6] == "000000"
)
print("Part One:", p1)
print("Part Two:", p2)

File: 2015/04/solution.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
)

func MineAdventCoin(secret []byte, num_zeros int, start int64) int64 {
	i := start
	for {
		hash := md5.Sum(strconv.AppendInt(secret, i, 10))
		encoded := hex.EncodeToString(hash[:])
		if strings.HasPrefix(encoded, strings.Repeat("0", num_zeros)) {
			return i
		}
		i++
	}
}

func main() {
	var secret []byte
	fmt.Scanln(&secret)
	p1 := MineAdventCoin(secret, 5, 1)
	p2 := MineAdventCoin(secret, 6, p1)
	fmt.Println("Part One:", p1)
	fmt.Println("Part Two:", p2)
}
